File: hw2/ShortCircuit.java
/* Kay Sweebe
 * CS471 - S Cooper
 * Programming2 - Short Circuit
 *
 * Demonstrate whether the Java programming language utilizes the concept of
 * short circuiting.
 *
 * Short circuiting occurs when only part of an expression needs to be
 * evaluated before finding the value.
 * See the following rules:
 *  if a and b and not a then false
 *  if a or  b and a     then true
 */

public class ShortCircuit {
  /** Checks to see if java includes short circuiting of compound statement
   **/

  static boolean test_short_circuit(){
    /** Helper function
     *  If contents get printed, then short circuiting does not happen
     *  Else short circuiting occurs
     *
     *  @return false
     **/
    System.out.println("I just got executed");
    return false;
  }// end test_short_circuit

  public static void main(String[] args){
    boolean x = false; // First condition should fail for and
    boolean y = true; // First condition should pass for or

    System.out.println("Start short circuit test AND...");
    if( x && test_short_circuit() );
    System.out.println("...End short circuit test");

    System.out.println("Start short circuit test OR...");
    if( y || test_short_circuit() );
    System.out.println("...End short circuit test");

    System.out.println("Helper function never prints");
    System.out.println("Short circuiting is used in Java");
  }// end main
}// end class

File: hw2/ShortCircuit.c
/** K. Sweebe
 *  CS 471 - S. Cooper
 *  Programming2 - Short Circuit
 *
 *  Demonstrate whether the C programming language utilizes the concept of
 *  short circuiting.
 *
 *  Short circuiting occurs when only part of an expression needs to be
 *  evaluated before finding the value.
 *  See the following logical rules:
 *      if a and b and not a then false
 *      if a or  b and a     then true
 */
#include <stdio.h>

int test_short_circuit();

int main()
{

    printf("Start short circuit test AND...\n");
    if( 0 && test_short_circuit() ) // first condition fails
        ;
    printf("End short circuit test...\n");

    printf("Start short circuit test OR...\n");
    if( 1 || test_short_circuit() ) // first condition passes
        ;
    printf("...End short circuit test\n");

    printf("Helper function never prints\n");
    printf("Short circuiting is used in C\n");
}// end main

int test_short_circuit()
{
    /** Helper function
     *  If contents get printed, then short circuiting does not happen
     *  Else short circuiting occurs
     */
    printf("I just got executed");
    return 0;

}//end test_short_circuit
